### PR TITLE
React: add typed target for change events

### PR DESCRIPTION
**Pull request by pzavolinsky**
Note: this PR is a proper implementation of #13665 and is also related to #12239.

---

When handling an `onChange` event it should be possible to access the changed value through `e.target.value` (see [this MDN example](https://developer.mozilla.org/en-US/docs/Web/Events/change#Example_Change_event_on_a_select)).

The following example (adapted from [React's documentation](https://facebook.github.io/react/docs/forms.html#the-textarea-tag)) illustrates the scenario:

```js
class EssayForm extends React.Component<{}, { value: string }> {
  constructor(props:{}) {
    super(props);
    this.state = {
      value: 'Please write an essay about your favorite DOM element.'
    };
  }
  render() {
    return <textarea
      value={this.state.value}
      onChange={e => this.setState({value: e.target.value})}
    />;
  }
}
```

This currently fails in the setState line with:
> error TS2339: Property 'value' does not exist on type 'EventTarget'.

---

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [x] Run `tsc` without errors.
- [ ] Run `npm run lint package-name` if a `tslint.json` is present.

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://facebook.github.io/react/docs/forms.html#the-textarea-tag
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "../tslint.json" }`.

### CODE REVIEW

**Comment by pzavolinsky** [comment]
The build failed in `react-redux`. If I `git checkout upstream/master` and run:

**Comment by pzavolinsky** [comment]
Sorry the broken comment before.

Here is the proper test:

```shell
git checkout -b is-master-working upstream/master 
echo >> react/index.d.ts 
git add .
git commit -m 'test'
node node_modules/types-publisher/bin/tester/test.js
```

> Error in react-redux
> Command failed: node /home/user/src/DefinitelyTyped/node_modules/types-publisher/bin/tester/../../typescript-installs/2.0/node_modules/typescript/lib/tsc.js
> ../react-router/lib/IndexRedirect.d.ts(12,16): error TS2305: Module '"/home/user/src/DefinitelyTyped/history/index"' has no exported member 'Query'.
> ../react-router/lib/Redirect.d.ts(14,19): error TS2305: Module '"/home/user/src/DefinitelyTyped/history/index"' has no exported member 'Query'.
> ../react-router/lib/Router.d.ts(4,5): error TS2305: Module '"/home/user/src/DefinitelyTyped/history/index"' has no exported member 'QueryString'.
> ../react-router/lib/Router.d.ts(4,18): error TS2305: Module '"/home/user/src/DefinitelyTyped/history/index"' has no exported member 'Query'.
> ...

Can anyone help me troubleshoot this? Should I wait until this is fixed in `master`?

**Comment by pzavolinsky** [comment]
I've just created issue #14056 that, I believe, is related to the fact that forking `master` on a passing build sometimes has broken tests.

I'm going to rebase my fork to the latest upstream `master` but I'm still getting those `react-redux` issues so no hope for a passing build in this PR.

When the `react-redux` issues get fixed in `master` I'll rebase again and maybe we'll be able to merge this PR.

**Comment by pzavolinsky** [comment]
Build failure related to #14049

**Comment by mhegazy** [comment]
@pspeter3 @vsiao @johnnyreilly  and @RyanCavanaugh can you please review?

**Comment by johnnyreilly** [comment]
Given the controversy this has presented in the past I'm not sure whether to merge or not. I'd like to merge. But that seems to make some people terribly sad...

**Comment by pzavolinsky** [comment]
Hi @johnnyreilly,

> But that seems to make some people terribly sad...

Both [React docs](https://facebook.github.io/react/docs/forms.html#the-textarea-tag) and [MDN docs](https://developer.mozilla.org/en-US/docs/Web/Events/change#Example_Change_event_on_a_select) seem to agree that using `target` for the `onChange` handler is safe.

Even in the controversial PR #12239 the same person saying this:

> I've no idea why so many people use e.target (including me in the past), introducing stupid bugs where I try to get attributes from the wrong element

Later makes the following clarification:

> (...) e.target is only used in the special case of input onChange, which is guaranteed to come from the input. So no issue there.

If you are not sure about merging this PR just yet maybe we could ask for some additional review?

Any feedback is appreciated, and if I need to change anything in this PR just let me know.

**Comment by johnnyreilly** [comment]
> e.target is only used in the special case of input onChange, which is guaranteed to come from the input. So no issue there.

I had not spotted that.  In that case I'm happy for it to go in.  However, please can you fix the tests first?

**Comment by pzavolinsky** [comment]
@johnnyreilly unfortunately I inherited those broken tests from `master` and they don't have anything to do with my change (moreover, I'm not sure how to fix the broken build). 

I created issue #14080, maybe you can take a look at that?

Once that gets fixed in `master` I can rebase my fork and the tests will pass.

**Comment by johnnyreilly** [comment]
Hi @pzavolinsky,

I'm afraid I'm not involved with react-redux.  If you submit a PR that resolves that issue and resupplies this change I can take a look though.  

John

**Comment by pzavolinsky** [comment]
Yay, it was a long journey but the build finally passed!

@johnnyreilly do you think we are ready to merge?

**Comment by johnnyreilly** [comment]
If no-one objects by the end of the day then ping me again and I'll merge.

**Review by hinell — APPROVED** [review]
Looks fine.

**Comment by pzavolinsky** [comment]
Hi @johnnyreilly , is it merge time?

**Comment by johnnyreilly** [comment]
yeah it is 😉

**Comment by pzavolinsky** [comment]
Cool, thanks!

**Comment by bbenezech** [comment]
Please revert this PR.

Even if they only originate from childless inputs/textarea or select with restricted children set that do not originate events (`options`), **change events do bubble**.

``` typescript
render() {
  return <div onChange={e => {
        type CurrentTargetType = typeof e.currentTarget
        type TargetType = typeof e.target
        // CurrentTargetType => EventTarget & HTMLDivElement: INDEED
        // TargetType => EventTarget & HTMLDivElement: NOPE, it is either an EventTarget & HTMLInputElement or an EventTarget & HTMLSelectElement, but nobody knows at compile time
      }}>
      <input type="checkbox" value="some-value" />
      <select><options...></select>
   </button>
}
```

`e.target` is NOT the droid you are looking for. `e.currentTarget` is.

'Fixing' the semantics of the DOM API with broken TS typings is a poor idea, please use the DOM API as it is intended to, not as you think it should be.

This PR is not helping beginners, keeping the confusion between both. `target` and `currentTarget` may be the same at runtime for change events, but it is not an obligation.

Someone please add me as a maintainer or whatever or just ping me before merging target stuff.
Even if this PR is not as harmful has the ones before, it is still wrong.

When typing is out of the scope of the compiler, use type casting.

tldr; use `e.currentTarget`

**Comment by bbenezech** [comment]
Someone please help me @huan086 @dsifford

**Comment by pzavolinsky** [comment]
Hi @bbenezech , point taken about `onChange` event bubbling.

Personally, I still consider valuable to have typed `onChange` targets when the `target` is guaranteed to be the correct one.

I believe PR #14188 is a good compromise, preventing misuse of `target` when bubbling but still maintaining compatibility with (pre TS2) working code.

The changes in that PR are a bit bigger than this one, so the more people reviewing it the better.

Please let me know if that PR breaks anything and I'll try my best to fix it.

Cheers

**Comment by johnnyreilly** [comment]
@bbenezech @pzavolinsky  you are both now listed as maintainers of the React type definitions.  You should accordingly receive notifications.

For my money this is one that's going to run and run because the official documentation appears to only ever mention `target` and not `currentTarget`. https://facebook.github.io/react/docs/forms.html#controlled-components

So @bbenezech it might be worth asking them to change the docs to something more correct?  For myself I don't have a strong opinion on this.  As a user of react I don't like the casting that I inevitably do with `target` at present.  Perhaps I should be using `currentTarget` but given the docs never seem to advocate this I haven't.  On the other hand, assuming the bubbling scenario is just as explained then I can see how that's an issue.

In the end I don't feel too strongly either way.  What I think would be beneficial for all is:

1. people coming to some kind of consensus of what's correct
2. someone contacting the react docs folks to encourage them to correct their documentation (if that's what's necessary to encourage people away from bad practice)
3. fixing the type definition one way or the other
4. a comment in the type definition to avoid people changing this again under pain of death.

I'm happy to merge changes that have been discussed and agreed.  This original change was only merged after what seemed like agreement. 

Let the discussion begin on  #14188 I guess.

**Comment by huan086** [comment]
Ops I've been mentioned :P

PR looks good to me, since it specifically targets change event, which I think only get raised by input and textarea, which is the special case I mentioned earlier

**Comment by dsifford** [comment]
Agree with @bbenezech 

I've been burned by `e.target` so many times. Not sure if the argument in this thread is solely about `onChange` events, but I just wanted to also mention that when I was newer with react I mostly got burned by this issue with `onClick` events.

Here's an example of one such issue that might get people:

https://jsfiddle.net/dsifford/9mju2Lmd/

**Pull request merged**
